### ISSUE TITLE
Возможность иметь `$ifdef` в `PABCSystem`, чтобы убирать из установщика для XP более новые типы .Net

### ISSUE THREAD

**Issue by SunSerega** [pascalabcnet/pascalabcnet#2983]
Для WinXP уже собирается отдельный установщик.
Перед его создание можно компилировать `PABCSystem` (или `RebuildStandartModules` целиком) с дополнительным параметром `/Define`:
```
..\bin\pabcnetc RebuildStandartModules.pas /rebuild /Define:.Net4.0
```
Это уже было нужно ранее, из за `ValueTuple`. А теперь ещё раз из за `IReadOnlyCollection<T>` в #2977.
Я понимаю что сильное развитвление создавать не стоит, но в случаях когда изменения с `$ifdef` влияют только на типы, не доступные из под WinXP - ничего плохого в этом нет.

**Comment by samuraiGH** [comment]
Всё это очень проблемное место языка сейчас. По хорошему всю совместимость нужно делать основываясь на директиве `{$TargetFramework}`.  Сейчас у нас полная неразбериха, для какой версии собираются бинарники. Допустим, атрибут `[Cache]` создаёт ключи из `ValueTuple`, которого нет в 4.0. Что делать, если разработка идёт на десятке, а надо запускать на XP -- не использовать кешированые функции.

Надо всё это комплексно решать, а не обмазывать условной компиляцией часть кусков, которые выстрелить могут

**Comment by SunSerega** [comment]
Боюсь "комплексно решать" в данном случае можно максимум обмазывая условной компиляцией (будь то `$ifdef`, или `if` в коде компилятора) отдельные куски кода которые могут выстрелить.
Или, можно подождать когда школы перейдут на линукс и наконец забыть об WinXP раз и на всегда...

**Comment by miks1965** [comment]
Не понял про ifdef. Это не сработает. ifdef вырезает куски кода на уровне лексера. А PABCSystem хранится в откомпилированном виде. Невозможно сделать так чтобы по ifdef PABCSystem  докомпилировался под конкретную платформу. Надо иметь две версии PABCSystem. 

Issue закрывать?

**Comment by SunSerega** [comment]
> Невозможно сделать так чтобы по ifdef PABCSystem докомпилировался под конкретную платформу.

Возможно. Вы уже делаете отдельный установщик для WinXP. В него можно упаковать свой .pcu файл. А в IDE анализатор кода от .pas файла с `$ifdef` не сломается. У меня они по всюду в `OpenCLABC` (для дебаг-версии) - их адекватно обрабатывает.

**Comment by miks1965** [comment]
На компьютере, на котором компилируется pcu, не установлен Windows XP. Поэтому условия типа $ifdef WindowsXP не пройдут. Они не будут выполняться.

**Comment by SunSerega** [comment]
Посмотрите на строчку кода в заголовке: Одна демонстрирует как можно запустить компиляцию `RebuildStandartModules.pas` перед созданием установщика для WinXP.
В самом конце команды стоит `/Define:.Net4.0` - это то, что будет тригерить `$ifdef`, заставляя `PABCSystem.pas` создавать особый .pcu файл для установщика для WinXP.

**Comment by miks1965** [comment]
Ну всё равно в коде PABCSystem будут эти ifdef - может что то сломаться у интеллисенса и при переходе по ошибкам в стандартные модули

**Comment by SunSerega** [comment]
> может что то сломаться у интеллисенса

И про это я тоже в начале сказал: У меня давно в `OpenCLABC` куча `$ifdef`-ов, я с помощью них контролирую уровень дебага. В том что от `OpenCLABC` попадает в установщик - весь дебаг вырублен. Анализатор кода с этим работает. Переход к определению и к ошибкам тоже работает. У меня даже повода issue создать не было - всё так идеально работает уже.

**Comment by SunSerega** [comment]
> Но я пока не понимаю, как компилировать PabcSystem под Net 4/0

Да, это тоже проблема - отсутствие возможности тестировать сразу...
Надо бы посмотреть ещё на GitHub Actions - проверить можно ли там выбрать версию .Net, чтобы проверять собирается ли сразу, а не когда у кого-то из XP-шников упадёт компилятор.

**Comment by miks1965** [comment]
Надо во первых это сделать локально. Во вторых, не нужны всякие ifdef секции - это другая проблема, которую можно решить просто - стандартный модуль надо писать в терминах NET 4.0.

**Comment by SunSerega** [comment]
Локально только с помощью докеров.

А если в стандартном модуле только `.NET 4.0` - этот функционал потом никак нельзя расширить, потому что типы уже запечатаны.
Писать только то что будет совместимо с версией, которая уже много лет не поддерживается её же авторами - это не решение. Это копание могили под своими же ногами.

**Comment by miks1965** [comment]
Ну до сего момента вас всё устраивало. И потом там нет функционала для использования новых возможностей

**Comment by SunSerega** [comment]
> Ну до сего момента вас всё устраивало

Это к чему вообще довод? Я создал issue когда появилась осязаемая причина. Если б я создал её раньше - вы бы закрыли не думая потому что нет хороших примеров.

> И потом там нет функционала для использования новых возможностей

В смысле нету? В данном случае вопрос в `IReadOnlyCollection<T>` и всех `ValueTuple`. Их вполне возможно использовать, если не из под WinXP.
А в WinXP я уже показал что такое легко отрубается и ничем не мешает - а вы продолжаете "а давайте не надо ничего менять".

**Comment by miks1965** [comment]
Я не вижу этой осязаемой причины - IReadOnlyCollection<T> - это придумали вы и наверное только вам это нужно. Тип диапазона - он вспомогательный, и его можно реализовывать без всего этого.

**Comment by samuraiGH** [comment]
>Локально только с помощью докеров.

В данный момент -- возможно и так

Но как я и писал выше, надо учить компилятор понимать для какой версии он что-либо собирает
Модули должны хранить свой собственный `{$TargetFramework}`
Компилятор должен на него смотреть и не собирать, если в модуле используются типы, которых в указанной версии нет

**Comment by samuraiGH** [comment]
В идеале должно быть 2 модуля `PABCSystem`. Один с {$TargetFramework net4.0} другой допустим `{$TargetFramework 4.8}` или 4.7 -- не суть

В полную версию системы нужно поставлять оба модуля. Если в основной программе стоит какой-нибудь древний {$TargetFramework}, то и PABCSystem подключать древний. В версию для WinXP поставлять только `PABCSystem` с таргетом на 4.0

Я позже сформирую issue, в которой подробно распишу как всё это сделать по хорошему надо

Что касается варианта в шапке. Я согласен, что очень неудобно при изменениях работать с оглядкой на .net 4.0. Сам хотел недавно вставить `IReadOnlyCollection` в `MatrSlice`. Однако, когда я пишу код, я более менее уверен, что всё это запустится на любой версии фреймворка. Чем больше типов из высоких версий вот так без предупреждения добавляется в стандартную библиотеку -- тем меньше такой уверенности. Учитывая, что я пишу программы, которые нет нет, да и используют другие люди, для меня проблема вот этих непоняток с версией более чем актуальна

**Comment by miks1965** [comment]
Если вы пишете для себя, то нет проблем - пишите без оглядки. Стандартный системный модуль на то и стандартный, что не должен обеспечивать всех этих тонкостей. Я жалею, что принял этот коммит, в котором исправлялся Range в системном модуле ради мнимого ускорения. Стандартный системный модуль устоялся и не требует новых возможностей следующих версий NET. Тем более, что всё ограничивается Framework 4.8 - которая уже тоже сильно устарела.

Обёртки над всем новым можно делать в других стандартных модулях, которые не будут работать под .NET 4.0 - что правильно.

Еще раз - абсолютно не понимаю предложения в заголовке топика - зачем портить стандартный модуль хотелками одного человека, которыми никто не будет пользоваться

**Comment by samuraiGH** [comment]
>хотелками одного человека, которыми никто не будет пользоваться

Тут не соглашусь. Реализовывать те интерфейсы, которые действительно подходят это нормальная и необходимая практика для распространяемых модулей
>И - зачем поставлять два стандартных модуля на компьютере с WinXP - ума не приложу.

Тут Вы наверно невнимательно прочитали моё сообщение. Там наоборот написано, что на XP -- только старый вариант, в обычный установщик -- оба варианта

**Comment by miks1965** [comment]
Хорошо. Давайте чётко.

Итак, что вы КОНКРЕТНО хотели бы сделать в PABCSystem что по разному бы работало под Windows XP и под старшими Windows?

**Comment by SunSerega** [comment]
Изначальная идея была абстрактная - чтобы не застревать в прошлом.

Мне `IReadOnlyCollection<T>` в списке реализуемых интерфейсов `IntRange` ничем не понадобилось, но как и сказал @samuraiGH - это нормально, реализовывать все подходящие интерфейсы.

Затем обнаружилось что `IReadOnlyCollection<T>` не доступен из под WinXP. Принципиально против поддержки WinXP я ничего не имею. Но никогда ничего не менять чтобы ничего не сломалось - это таки не решение.

По крайней мере мне ваша позиция звучала как "никогда ничего не менять", но

> Обёртки над всем новым можно делать в других стандартных модулях, которые не будут работать под .NET 4.0 - что правильно.

Я с этим согласен. В этом случае получается это я думал о решении с минимумом усилий:

Если сделать `$ifdef` в `PABCSystem` - получится что поддерживаем только 1 модуль. Никогда не будет так что `PABCSystem` для WinXP ушёл вперёд относительно более новой версии для других OS. И несовместимость с новой версией всегда будет минимальной, потому что в `$ifdef` только те части, которые не доступны из под WinXP.

Но если вы думаете скорее в сторону вообще не трогать текущий `PABCSystem`, подождать пока будет возможность использовать паскаль с .Net Core и сделать там всё новое - это направление я поддерживаю. Всё равно в .Net Core многое переделано/расширено, особенно что касается интерфейсов. А значит значительная часть текущего `PABCSystem` будет сильно устаревшей после перехода на .Net Core.

**Comment by miks1965** [comment]
Это main.pas
```
uses a;
{$define x}
begin
end.
```
Это a.pas
```
unit A;
{$ifdef x}
var a1 := 2;
{$else}
x = 3;
{$endif}
end.
```

Символ x в модуле a не определен. Дело в том, что в Delphi define распространяется на одну единицу компиляции. 

Это делает невозможным использование ifdef в PABCSystem.

**Comment by SunSerega** [comment]
> Это делает невозможным использование ifdef в PABCSystem.

Нет, это делает невозможным контроль символов `PABCSystem` из подключающей его программы. То есть никак нельзя из вне поменять что доступно в интерфейсной части `PABCSystem`.

Это так для любых модулей, но обычные хотя бы может пере-компилировать в разных ситуациях, но стандартные модули вообще не перекомпилирует - вместо этого всегда использует готовый .pcu файл, в который попадает результат прохода пре-компилятора по всем `$ifdef` ещё до создания установщиков.

Но это вы вообще вернулись куда то не туда. О таком контроле в этой issue речь не шла ни в один момент...
Попробуйте перечитать обсуждение. А то я уже несколько раз это повторил в разной форме, а сегодня вы то ли снова забыли, то ли так и не до-разобрались...

Ближайшее о чём тут была речь - это предложение @samuraiGH чтобы в зависимости от `$TargetFramework` выбирались некоторые вещи в компиляторе, как версия `PABCSystem` и принцип работы `[Cache]`.
Потому что сейчас если откомпилировать программу с `[Cache]` на более новой ОС, а потом попытаться запустить сгенерированный .exe на WinXP - этот .exe крашнется.

**Comment by miks1965** [comment]
Насколько я понял из обсуждения, требуется по каким то причинам иметь две версии PABCSystem для разных систем.

Это небесполезная вещь - когда мы перейдем на .NET. то это будет востребовано.

Переименуйте пожалуйста топик согласно этой идее. Заголовок топика сбивает, предлагая конкретную неправильную реализацию этой проблемы - в PABCSystem всякие ifdef в том смысле, как они понимаются сейчас, бесполезны и не решают задачу.

Переименуйте топик

**Comment by SunSerega** [comment]
Изначальная issue с её названием была напоминанием самому себе, но с вашим направлением решения я таки согласен, поэтому переименовал.

**Comment by miks1965** [comment]
А что такое легаси? И там правильно слова согласованы в заголовке?

**Comment by SunSerega** [comment]
"legacy" переводится как "наследие", но для программистов имеет особый смысл: Это старый код, который на текущий день нереально поменять/улучшить.

Насчёт согласования, вы имеете в виду "версий" vs "версии"? Мне кажется я таки правильную форму тут использовал, для того что я имел в виду, но с вашим вопросом я теперь менее уверен...

**Comment by ibond84** [comment]
Ну RevisionClass.Revision сохранять в PCU. Онa увеличивается при вызове _GenerateAllSetups.bat

**Comment by samuraiGH** [comment]
@ibond84 это точно к этой issue относится?
